quipucords/api/fact/util.py

```diff
@@ -0,0 +1,183 @@
+    :returns: dict with bool indicating error and result.
+    """
+    if not fact_collection_json.get(SOURCES_KEY):
+        return {ERRORS_KEY: True,
+                RESULT_KEY: {
+                    SOURCES_KEY: _(messages.FC_REQUIRED_ATTRIBUTE)}}
+
+    return _validate_sources_json(fact_collection_json.get(SOURCES_KEY))
+
+
+def _validate_sources_json(sources_json):
+    """Validate sources field.
+
+    :param sources_json: list of sources.  Each source is a dict.
+    :returns: dict with 2 lists.  Valid and invalid sources.
+    """
+    valid_sources = []
+    invalid_sources = []
+    for source_json in sources_json:
+        result = _validate_source_json(source_json)
+        if result[ERRORS_KEY]:
+            invalid_sources.append(result[RESULT_KEY])
+        else:
+            valid_sources.append(source_json)
+
+    has_errors = len(invalid_sources) > 0
+    return {ERRORS_KEY: has_errors,
+            RESULT_KEY: {
+                VALID_SOURCES_KEY: valid_sources,
+                INVALID_SOURCES_KEY: invalid_sources}}
+
+
+def _validate_source_json(source_json):
+    """Validate source fields.
+
+    :param source_json: The dict representing facts for a source
+    :returns: None if no errors or a JSON dict with the errors
+    """
+    invalid_field_obj = {}
+    source_id = source_json.get(SOURCE_ID_KEY)
+    has_error = False
+    if not source_id:
+        has_error = True
+        invalid_field_obj[SOURCE_ID_KEY] = _(
+            messages.FC_REQUIRED_ATTRIBUTE)
+
+    if not has_error and not isinstance(source_id, int):
+        has_error = True
+        invalid_field_obj[SOURCE_ID_KEY] = _(
+            messages.FC_SOURCE_ID_NOT_INT)
+
+    if not has_error:
+        source = Source.objects.filter(pk=source_id).first()
+        if not source:
+            has_error = True
+            invalid_field_obj[SOURCE_ID_KEY] = _(
+                messages.FC_SOURCE_NOT_FOUND % source_id)
+
+    source_type = source_json.get(SOURCE_TYPE_KEY)
+    if not source_type:
+        has_error = True
+        invalid_field_obj[SOURCE_TYPE_KEY] = _(
+            messages.FC_REQUIRED_ATTRIBUTE)
+
+    if not has_error and not \
+            [valid_type for valid_type in Source.SOURCE_TYPE_CHOICES
+             if valid_type[0] == source_type]:
+        has_error = True
+        valid_choices = ', '.join(
+            [valid_type[0] for valid_type in Source.SOURCE_TYPE_CHOICES])
+        invalid_field_obj[SOURCE_TYPE_KEY] = _(
+            messages.FC_MUST_BE_ONE_OF % valid_choices)
+
+    facts = source_json.get(FACTS_KEY)
+    if not facts:
+        has_error = True
+        invalid_field_obj[FACTS_KEY] = _(
+            messages.FC_REQUIRED_ATTRIBUTE)
+
+    result = {ERRORS_KEY: has_error}
+    if has_error:
+        error_json = {}
+        error_json[SOURCE_KEY] = source_json
+        error_json[ERRORS_KEY] = invalid_field_obj
+        result[RESULT_KEY] = error_json
+    return result
```

Comment:
Couldn't we just return a tuple of the results and errors instead of a dictionary with keys?

Reply:
👍